design/one-pager-simple-class-selection.md

```diff
@@ -0,0 +1,411 @@
+      region: us-west
+      replicated: "true"
+```
+
+These resource classes specify that they can satisfy claims for replicated Redis
+3.2 instances on the west coast of the USA via their labels:
+
+```yaml
+apiVersion: cache.gcp.crossplane.io/v1alpha2
+kind: CloudMemorystoreInstanceClass
+metadata:
+  name: gcp-dev
+  labels:
+    engineVersion: "3.2"
+    region: us-west
+    replicated: "true"
+specTemplate:
+  tier: STANDARD_HA
+  region: us-west2
+  memorySizeGb: 1
+  redisVersion: REDIS_3_2
+---
+apiVersion: cache.azure.crossplane.io/v1alpha2
+kind: RedisClass
+metadata:
+  name: azure-dev
+  labels:
+    engineVersion: "3.2"
+    region: us-west
+    replicated: "true"
+specTemplate:
+  resourceGroupName: group-westus-1
+  location: West US
+  sku:
+    name: Basic
+    family: C
+    capacity: 0
+```
+
+Crossplane could (at some future time) solve the specific example above by
+[teaching each resource claim controller] how to match standardised resource
+claim fields to their provider specific resource class fields, but this is true
+only for a very small class of fields that are both highly relevant to claim
+authors and highly likely to translate to every conceivable managed resource
+that could satisfy a claim. Geographic region, node count, and database size are
+examples of such fields.
+
+Label selectors approximate this functionality while it does not yet exist and
+will compliment it when it does. There will likely be characteristics of
+resource classes that claim authors want to match on that cannot be first class
+configurable fields of the resource claim `spec` because they do translate to
+all managed resources.
+
+### Unopinionated Resource Claims
+
+In the absence of spec based scheduling a resource claim that omits the
+`.spec.classSelector` field, as well as the `.spec.classRef` and
+`.spec.resourceRef` fields that allow them to explicitly specify a class or
+existing managed resource respectively will be deemed unopinionated about what
+resource class it needs.
+
+Kubernetes [states that]:
+
+> The semantics of empty or non-specified selectors are dependent on the
+> context, and API types that use selectors should document the validity and
+> meaning of them.
+
+This document proposes that Crossplane fall back to a default resource class - a
+class annotated (not labelled) `resourceclass.crossplane.io/is-default-class:
+"true"` - in order to satisfy an unopinionated resource claim. If there is no
+default resource class the resource claim will not be satisfied and will remain
+unbound. If there are multiple default resource classes one will be chosen at
+random.
+
+## Technical Implementation
+
+Under the hood, resource class selection for any resource class that specified a
+`classSelector` (and did not specify a `classRef` or `resourceRef`) would be
+implemented as a race between a series of "class scheduler" controllers.
+
+In Crossplane there is not one controller for each resource claim kind, but
```

Comment:
How about we allow managed resource CRDs to have label that they can tell what resource claim they can be used for? `CloudSQLInstance` CRD can have a label `crossplane.io/resource-claim: MySqlInstance`. One generic `MySqlInstance` controller could get all CRDs with that label, lists all instance of these CRDs and pick one randomly. Pretty much like kinds would register themselves to a portable resource claim kind as candidate.

Reply:
This is an intriguing idea. I like that it would allow us to assess all possible classes that could satisfy a claim in one place (instead of assessing each kind of class separately). If we did want to have classes effectively register support for a kind of claim, I think the CRD is a good place to do so.

Two potential issues I could see, though:

* If we did want to expand to making claim scheduling decisions by comparing the spec of the claim to the spec of the class we'd probably need whatever was doing that comparison to have access to the class's Go struct. It would probably be _possible_ to do this using `*unstructured.Unstructured`, but we'd be working with raw JSON (informed by the CRD schema) to make scheduling decisions.
* If/when we introduced non-portable claims like a `DynamoDb` claim, the scheduling/defaulting controllers would like in the stack where that claim was defined, while portable claims like `MySQLInstance` would have their scheduling/defaulting controllers defined in Crossplane runtime. It's probably a stretch to call this an "issue" - more of an asymmetry.

Reply:
> If we did want to expand to making claim scheduling decisions by comparing the spec of the claim to the spec of the class

I don't understand why we'd need to do that comparison. In my imagination, this controller's job would be just to choose a class and put its reference to `classRef` of `MySqlInstance` claim that it reconciles. The type of the class would be unknown but only thing needed for referencing is metadata anyway.

> It's probably a stretch to call this an "issue" - more of an asymmetry.

I agree. On the positive side, stack developers can implement their own unique scheduling mechanism for non-portable claims if they really wanted to.